planner/core/rule_aggregation_skew_rewrite.go

```diff
@@ -0,0 +1,262 @@
+			}
+
+			// now the distinct is not needed anymore
+			newAggFunc.HasDistinct = false
+			topAggFuncs = append(topAggFuncs, newAggFunc)
+		} else {
+			// only count() will be decomposed to sum() + count(), the others will keep same
+			// original aggregate functions will go to bottom aggregate without any change
+			bottomAggFuncs = append(bottomAggFuncs, newAggFunc)
+
+			// cast to Column, if failed, we know it is Constant, ignore the error message,
+			// we will later create a new schema column
+			aggCol, _ := newAggFunc.Args[0].(*expression.Column)
+
+			// firstrow() doesn't change the input value and type
+			if newAggFunc.Name != ast.AggFuncFirstRow {
+				aggCol = &expression.Column{
+					UniqueID: agg.ctx.GetSessionVars().AllocPlanColumnID(),
+					RetType:  newAggFunc.RetTp,
+				}
+			}
+			bottomAggSchema.Append(aggCol)
+
+			if newAggFunc.Name == ast.AggFuncCount {
+				cntIndexes = append(cntIndexes, i)
+				sumAggFunc, err := aggregation.NewAggFuncDesc(agg.ctx, ast.AggFuncSum,
+					[]expression.Expression{aggCol}, false)
+				if err != nil {
+					return nil
+				}
+				topAggFuncs = append(topAggFuncs, sumAggFunc)
+				topAggSchema.Columns[i] = &expression.Column{
+					UniqueID: agg.ctx.GetSessionVars().AllocPlanColumnID(),
+					RetType:  sumAggFunc.RetTp,
+				}
+			} else {
+				topAggFunc := aggFunc.Clone()
+				topAggFunc.Args = make([]expression.Expression, 0, len(aggFunc.Args))
+				topAggFunc.Args = append(topAggFunc.Args, aggCol)
+				topAggFuncs = append(topAggFuncs, topAggFunc)
+			}
+		}
+	}
+
+	// now create the bottom and top aggregate operators
+	bottomAgg := LogicalAggregation{
+		AggFuncs:     bottomAggFuncs,
+		GroupByItems: bottomAggGroupbyItems,
+		aggHints:     agg.aggHints,
+	}.Init(agg.ctx, agg.blockOffset)
+	bottomAgg.SetChildren(agg.children...)
+	bottomAgg.SetSchema(bottomAggSchema)
+
+	topAgg := LogicalAggregation{
+		AggFuncs:     topAggFuncs,
+		GroupByItems: agg.GroupByItems,
+		aggHints:     agg.aggHints,
+	}.Init(agg.ctx, agg.blockOffset)
+	topAgg.SetChildren(bottomAgg)
+	topAgg.SetSchema(topAggSchema)
+
+	if len(cntIndexes) == 0 {
+		appendSkewDistinctAggRewriteTraceStep(agg, topAgg, opt)
+		return topAgg
+	}
+
+	// it has count(), we have split it into sum()+count(), since sum() returns decimal
+	// we have to return a project operator that casts decimal to bigint
+	proj := LogicalProjection{
+		Exprs: make([]expression.Expression, 0, len(agg.AggFuncs)),
+	}.Init(agg.ctx, agg.blockOffset)
+	for _, column := range topAggSchema.Columns {
+		proj.Exprs = append(proj.Exprs, column)
+	}
+
+	// wrap sum() with cast function to keep output data type same
+	for _, index := range cntIndexes {
+		exprType := proj.Exprs[index].GetType()
+		targetType := agg.schema.Columns[index].GetType()
+		if !exprType.Equal(targetType) {
+			proj.Exprs[index] = expression.BuildCastFunction(agg.ctx, proj.Exprs[index], targetType)
+		}
+	}
+	proj.SetSchema(agg.schema.Clone())
```

Comment:
select a+b, c+d from foo;
here expressions are [a+b, c+d]
but schema is [column#5, column#6].

Reply:
When the expression has been changed, the schema should be changed at the same time?

Reply:
No, the top schema should be the same, the type and column id, otherwise, we have to change the parent operators reference to these columns all the way up.